### ISSUE TITLE
Map with long keys and :justify? true does not become justified

### ISSUE THREAD

**Issue by kennyjwilli** [kkinnear/zprint#31]
Using zprint 0.4.2:

```clojure
(zp/zprint
 '(def foo
    {:a                                                      (abcdef-in12345678912)
     :abcdef-abcdefgh-abc-abcdefghi-abcdefgh-abcdef-abcdefgh false})
 80
 {:map {:justify? true
        :comma?   false}})

(def foo
  {:a (abcdef-in12345678912)
   :abcdef-abcdefgh-abc-abcdefghi-abcdefgh-abcdef-abcdefgh false})
=> nil
```

When it should be:

```clojure
(def foo
  {:a                                                      (abcdef-in12345678912)
   :abcdef-abcdefgh-abc-abcdefghi-abcdefgh-abcdef-abcdefgh false})
```

**Comment by kkinnear** [comment]
Thanks for noticing this!

When I try your example, it takes 81 characters to print it justified, and so it gives up and won't do the justification in 80 characters (as it knows about the right paren after the (abcdef...)).  When I do the line width on your "When it should be:", I also get 81 characters -- the ")" is in the 81st column.  I worked really hard to account for those pesky parentheses, and if the justification will go too far, it falls back to non-justified.

Unfortunately, I apparently didn't work hard enough, because if I change the 80 to 81 it still won't print justified, and requires the size to be 82 to print.  Sigh. I'm guessing that this has something to do with not accounting for the not having the commas that you didn't want.  If I let it do the commas, it will justify in 82 characters.  I'll see what I can do about getting it to work with no commas.

Thanks for bringing this to my attention.  Getting the actual size right, and printing everything that will fit in whatever size you specify accurately is one of the reasons that I wrote zprint in the first place.  I have lots of tests for just this sort of situation, but I don't think I did much w/out commas, so I'll look there first.

**Comment by kkinnear** [comment]
I fixed this in 0.4.3, so now the comma's are accounted for correctly -- if you didn't want commas, they not only don't appear (which already worked), but we don't include them in the count (which didn't work).  Thanks for bringing this to my attention, though I understand that this wasn't exactly what you were concerned about.

**Comment by kennyjwilli** [comment]
So does a map become justified now?

**Comment by kkinnear** [comment]
The example you gave wasn't justified because you gave it a width of 80, and if it were to be justified, the right parenthesis after the ...8912 would be in column 81.  If you change the 80 in your example to 81, then (in zprint 0.4.3, but not in zprint 0.4.2) the map will format justified.  The justification is only performed if it will actually fit within the width, and if it doesn't (as it doesn't if the width is 80), then it falls back to "normal" formatting.  If the pairs don't fit on the same line, the the justification is ... not really justification, so that is why it falls back for normal formatting if the pairs don't fit on a single line.

The thing I fixed in 0.4.3 was that, in 0.4.2, your example wouldn't justify with a width of 81 either, it needed 82 to properly justify your example.  That's because it was counting the comma, even though you told it there wasn't a comma.  Now in 0.4.3, if you turn off commas, it doesn't count them in the line length when it tries to see if the line will fit in the width.  So, in 0.4.3, it will justify if you change the 80 in your example to 81.

Please let me know if this doesn't work, it is possible I made some mistake in testing your example.  Thanks again for bringing this up!

As an aside, if you want to see a justified map, try `(czprint nil :explain-justified)`  This is the regular `(czprint nil :explain)` options map output, justified.  It used to look better before I added `additional-libraries` at the start of it, but now that shifts things over to the right a lot.  Poor choice of key, but I can't change it now, unfortunately.
